packages/htmljs/visitors.js

```diff
@@ -34,85 +35,89 @@ Visitor.def = function (options) {
   };
-  subType.prototype = new curType;
-  subType.extend = curType.extend;
-  subType.def = curType.def;
-  if (options)
-    _assign(subType.prototype, options);
+  subType.prototype = new CurType();
+  subType.extend = CurType.extend;
+  subType.def = CurType.def;
+  if (options) _assign(subType.prototype, options);
   return subType;
 };
 
 Visitor.def({
-  visit: function (content/*, ...*/) {
-    if (content == null)
+  visit(content/* , ... */) {
+    if (content == null) {
       // null or undefined.
       return this.visitNull.apply(this, arguments);
+    }
 
     if (typeof content === 'object') {
       if (content.htmljsType) {
         switch (content.htmljsType) {
-        case Tag.htmljsType:
-          return this.visitTag.apply(this, arguments);
-        case CharRef.htmljsType:
-          return this.visitCharRef.apply(this, arguments);
-        case Comment.htmljsType:
-          return this.visitComment.apply(this, arguments);
-        case Raw.htmljsType:
-          return this.visitRaw.apply(this, arguments);
-        default:
-          throw new Error("Unknown htmljs type: " + content.htmljsType);
+          case Tag.htmljsType:
+            return this.visitTag.apply(this, arguments);
+          case CharRef.htmljsType:
+            return this.visitCharRef.apply(this, arguments);
+          case Comment.htmljsType:
+            return this.visitComment.apply(this, arguments);
+          case Raw.htmljsType:
+            return this.visitRaw.apply(this, arguments);
+          default:
+            throw new Error(`Unknown htmljs type: ${content.htmljsType}`);
         }
       }
 
-      if (isArray(content))
-        return this.visitArray.apply(this, arguments);
+      if (isArray(content)) return this.visitArray.apply(this, arguments);
 
       return this.visitObject.apply(this, arguments);
-
-    } else if ((typeof content === 'string') ||
-               (typeof content === 'boolean') ||
-               (typeof content === 'number')) {
+    }
+    if ((typeof content === 'string') ||
+      (typeof content === 'boolean') ||
+      (typeof content === 'number')) {
       return this.visitPrimitive.apply(this, arguments);
-
-    } else if (typeof content === 'function') {
+    }
+    if (typeof content === 'function') {
       return this.visitFunction.apply(this, arguments);
     }
 
-    throw new Error("Unexpected object in htmljs: " + content);
-
+    throw new Error(`Unexpected object in htmljs: ${content}`);
   },
-  visitNull: function (nullOrUndefined/*, ...*/) {},
-  visitPrimitive: function (stringBooleanOrNumber/*, ...*/) {},
-  visitArray: function (array/*, ...*/) {},
-  visitComment: function (comment/*, ...*/) {},
-  visitCharRef: function (charRef/*, ...*/) {},
-  visitRaw: function (raw/*, ...*/) {},
-  visitTag: function (tag/*, ...*/) {},
-  visitObject: function (obj/*, ...*/) {
-    throw new Error("Unexpected object in htmljs: " + obj);
+  visitNull(nullOrUndefined/* , ... */) {
```

Comment:
I know they are there but never used. However, I don't know why they were added or if there should be a future use for them. We should therefor move their variables inside the comments and add an error to indicate that this is not implemented:

```js
...

visitNull(/* nullOrUndefined , ... */) {
  throw new Error('Visitor.visitNull is not implemented');
},

...

```

If future usage raises these errors we can analyze them and find a way for an implementation

Reply:
They actually are used. The visitor could call any or all of these depending on what types are in the AST it is visiting. For example, line 53 above calls visitNull. Each type of  visitor (for example, the TransformingVisitor in this file) overrides the functions for the ast nodes that it is interested in. The functions for other node types then do nothing.

Reply:
Really hard to understand what it is really doing

Reply:
> Really hard to understand what it is doing

This is mostly due to lack of documentation of the internals and applies to most of the Blaze packages.

> Each type of visitor (for example, the TransformingVisitor in this file) overrides the functions for the ast nodes that it is interested in. The functions for other node types then do nothing by default.

So this is similar to what abstract classes with uninmplemented base methods do, right?

Reply:
Yes, they are like abstract classes